Add the ability to specify the default page

ZeroErrors opened the pull request:
~~**Depends on:** https://github.com/openDsh/dash/pull/34~~

## Change Log
- Added "Default Page" setting

## Description
In some cases its useful to be able to display a different default page when first starting the app instead of being on a blank screen waiting for a device to connect.

### Future Work
Additional features like automatically swapping to the Android Auto page when a device is connected and automatically swapping back to the default when disconnected would also greatly enhance this feature. This type of feature could also be expanded for handling things like automatically displaying a reversing camera.

## Checklist:
  - [x] The code change is tested and works locally.

ZeroErrors committed:
Add the ability to specify the default page

icecube45 commented:
Not related to this PR - but just taking a second to plug the [Open Dash slack](https://join.slack.com/t/opendsh/shared_invite/zt-k9nh1v8j-0FhBx90rkaGCJ_rjCAeI8A)

I think the last invite link we had somewhere has expired at this point

icecube45 commented:
I like this setting
Only two suggestions:

- Probably a good idea to remove undisplayed pages from the default page options (this seems to default to AA if you select a page you turned off)
- Maybe the above is best done through a radio button next to the current checkboxes on turning pages on/off

rsjudka commented:
i agree with @icecube45, we'll probs want to better "merge" the 2 settings together (default page and enabled pages) mainly because changes to the enabled pages can affect the default page)

qt checkboxes have a "tristate" option... maybe that might be something to consider? it could be something like off, on, and default page

icecube45 commented:
Tristate appears to just track changes in box status - I think we'd need another row of exclusive checkboxes

Then again the QT Documentation isn't great - it may work for our purposes.

rsjudka commented:
oh yeah its definitely not its intended use haha 

but we can really use that 3rd state for whatever
![image](https://user-images.githubusercontent.com/16892670/102704682-45baa000-4244-11eb-8638-617044417f97.png)

rsjudka commented:
the way im imagining it is you take the `stateChanged` signal and do it one of two ways:

- if current state is off, set next state to on... if current state is on, set to off (if a default page is set) or "partial" (if a default page wasnt set yet)
- set the next state, and if the next state happens to be "partial", then find the previous default page and set that to just the on state

i would honestly need to code that up to see how intuitive it is from a UI perspective tho (mainly trying to minimize the amount elements needed), but a radio button to the side is probably easier to add and either seems fine to me

icecube45 commented:
> Future Work
> 
> Additional features like automatically swapping to the Android Auto page when a device is connected and automatically swapping back to the default when disconnected would also greatly enhance this feature. This type of feature could also be expanded for handling things like automatically displaying a reversing camera.

In addition to reverse camera tab switching -
Android Auto / openauto should also be spitting out media and audio focus requests, good idea to make dash properly follow those, and change to AA tab during media focus requests.

ZeroErrors committed:
Rework pages settings

ZeroErrors commented:
So tried a different way of handling the pages config, let me know what you think.

![screenshot](https://user-images.githubusercontent.com/1109304/102754365-cac8c500-43bc-11eb-8f7f-fcc139deac97.png)


I also renamed it from 'default' page to 'home' page and rebased the branch on develop with the latest changes.

rsjudka commented:
i can see that working for like a desktop application, but for something that's touch based and often used on smaller screens we need to try to keep things minimal

With this, im assuming the user would need to tap on the page, and then select one of the actions to then perform on it? Even though it's only one additional tap, with screens that are less responsive/accurate, it might get annoying to mistap or not have your touch registered at all

I like @icecube45 suggestion of radio boxes, since it's only an additional element in the horizontal space and handling of a single selection is already implemented since they're in a group box

ZeroErrors committed:
Use radio buttons for home page setting

ZeroErrors commented:
Ok, replaced it with radio buttons.
![screenshot](https://user-images.githubusercontent.com/1109304/102838306-c6002180-444d-11eb-836b-88a77a0f9e98.png)

Something I was thinking about with the way I went was also the ability to reorder the pages and also to help deal with the expanding number of pages, but I guess they can be dealt with later when it becomes needed.

icecube45 commented:
That's a good point - in the future maybe we have to make a custom widget to make it (small) touchscreen compatible.

rsjudka commented:
I would say remove the icons, maybe right align the radio buttons? Not sure if that's better or not lol

Yeah I considered allowing the pages to be reordered but it didn't feel like something that many people cared for

I dont see too many pages being added anyway, but will probs set a limit on the number of pages that can be enabled at a time (likely 7 based on material design spec)

rsjudka commented:
@icecube45 could you explain that a bit more? Not really sure what your mean

icecube45 commented:
Like if we were to design a custom selector widget that would allow for defaulting and ordering, while remaining simplistic and easy to use on a touchscreen. I have some ideas in the back of my mind, but nothing fleshed out

rsjudka commented:
Oh gotcha

Let me know if you need help moving anything up in qt (or seeing if something is feasible 😅)

ZeroErrors commented:
Here is some screenshots of a few different layouts.
I have a concern that without icons there is no indication as to what the radio buttons mean and since there is no direct user feedback like the checkboxes, which instantly hide or show the pages in the sidebar, you would have no idea what it does until you restart dash.

**1. Both left with icons**
![left-icons](https://user-images.githubusercontent.com/1109304/102954616-cd91fa00-4523-11eb-961e-f660a2f181f7.png)

**2. Both right with icons**
![both-right-icons](https://user-images.githubusercontent.com/1109304/102955567-22367480-4526-11eb-93ec-1adfd27f93eb.png)

**3. Home right with icons**
![right-icons](https://user-images.githubusercontent.com/1109304/102954620-cec32700-4523-11eb-8437-f43793779daa.png)

**4. Both left without icons**
![left](https://user-images.githubusercontent.com/1109304/102954614-cc60cd00-4523-11eb-9115-8ae5e0f5c659.png)

**5. Both right without icons**
![both-right](https://user-images.githubusercontent.com/1109304/102955575-26629200-4526-11eb-88b3-a17786feed3f.png)

**6. Home right without icons**
![right](https://user-images.githubusercontent.com/1109304/102954619-ce2a9080-4523-11eb-9ed4-7e9145228bf9.png)

dannybarake commented:
I prefer option 3. This is most user friendly for newbies of Dash.

stefan-sherwood commented:
I would use on-off switches instead of checkboxes for consistency. I also think it would look cleaner if the default page was a separate -dropdown- arrow selector (don't know what it's called) that populated based on which pages were enabled.

(this is the selector control I'm referring to)
![image](https://user-images.githubusercontent.com/2094779/102999713-8db51d80-44f7-11eb-99fe-1bfe4b1570cb.png)

robert5974 commented:
I like Stefan's suggestion.  The selector makes sense.

icecube45 commented:
I'm more hesitant on the use of a selector - I'm not sure it's intuitive to a user as to what happens when you disable a page that's set as a default page.

stefan-sherwood commented:
In that case the default changes to a different item. All of this is probably rarely going to be changed so clarity of the state is arguably the most important thing. Regardless, how is the radio button any different in terms of intuitiveness of what happens when you disable the default item? It is functionally identical.

stefan-sherwood commented:
Also note that you can put a label next to it that says "Home page". That makes it pretty clear.

ZeroErrors committed:
Merge branch 'develop' into pages-default

UniCizin commented:
hey guys,

any news on this? would love to "autostart" into bluetooth media view as i am using my phone and spotify as source

rsjudka commented:
not sure where things left off here... i think a selector here would be just fine, and totally something i can pick up after i get through some of the pieces I'm currently working on
